proposals/metadata.md

```diff
@@ -0,0 +1,103 @@
+on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+express or implied. See the License for the specific language governing
+permissions and limitations under the License.
+-->
+
+### Container metadata feature
+In order to solve the issues related to container metadata like: [#288](https://github.com/aws/amazon-ecs-agent/issues/288) and [#456](https://github.com/aws/amazon-ecs-agent/issues/456), we are considering adding a new feature "Container Metadata" similar to [EC2 Instance Metadata Service](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html) in ECS agent to provide the container metadata.
+
+By adding this feature, you can query the information about the task, container and container instance from within the container or from the container instance by reading the metadata file for each container. The following information will be available from this feature:
+ * Information from Docker:
+  * Container Network Information
+  * Container ID from Docker
+  * Container Name
+  * Image Name
+  * Image ID
+  * Port Mapping
+ * Information from ECS:
+  * Cluster Arn
+  * Container Instance Arn
+  * Task Arn
+
+ECS agent will create a file for each container under the directory `/var/lib/ecs/data/` and then mount the file to the container under a directory eg: `/ecs/metadata` with read only permission when the container is being created. Also an environment variable denoting the file path eg: `ECS_CONTAINER_METADATA_FILE` will be injected into the container. The `ECS_CONTAINER_METADATA_FILE` is consisted of several parts like: "/ecs/metadata/container_name_from_task_definition/metadata.json". After the container is started, ECS agent will update the metadata of container into this file. Then the container can get its metada information by parsing the json file with path `ECS_CONTAINER_METADATA_FILE`. This directory will be removed when task is being clean up. Also this feature will be configurable from the environment variable or in the task definition, eg: adding a field `"metadataEnable": true`. The detailed steps when start a container will be:
+* Check whether the feature is enabled, if not then go directly create and start the container.
+* Check the task definition make sure the path `/ecs` is not mounted, if it's used in the task definition, then disable the feature.
+* Create metadata file for each container under `/var/lib/ecs/data/` for each container.
+* Create the container with the two extra steps:
+  * Mount the metadata file created above to a specific directory eg: `/ecs/metadata/` with read only permission for each container.
+  * Inject the metadata file path as an environment variable to the container eg: `ECS_CONTAINER_METADATA_FILE=/ecs/metadata/container1`.
+* After the container is created, write the basic metadata information into the metadata file.
+* Start the container and update the networking information into the metadata file.
```

Comment:
This issue is lessened if more files are used vs a single json file because polling/inotify come more naturally.

All the same, polling/inotify could be used to wait for the relevant keys to exist